lib/srv/db/proxy_test.go

```diff
@@ -0,0 +1,192 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package db
+
+import (
+	"context"
+	"io"
+	"net"
+	"testing"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/multiplexer"
+	"github.com/gravitational/teleport/lib/utils"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+)
+
+// TestProxyProtocolPostgres ensures that clients can successfully connect to a
+// Postgres database when Teleport is running behind a proxy that sends a proxy
+// line.
+func TestProxyProtocolPostgres(t *testing.T) {
+	ctx := context.Background()
+	testCtx := setupTestContext(ctx, t, withSelfHostedPostgres("postgres"))
+	t.Cleanup(func() { testCtx.Close() })
+	go testCtx.startHandlingConnections()
+
+	testCtx.createUserAndRole(ctx, t, "alice", "admin", []string{"postgres"}, []string{"postgres"})
+
+	// Point our proxy to the Teleport's db listener on the multiplexer.
+	proxy, err := newTestProxy(testCtx.mux.DB().Addr().String())
+	require.NoError(t, err)
+	t.Cleanup(func() { proxy.Close() })
+	go proxy.Serve()
+
+	// Connect to the proxy instead of directly to Postgres listener and make
+	// sure the connection succeeds.
+	psql, err := testCtx.postgresClientAddr(ctx, proxy.Address(), "alice", "postgres", "postgres", "postgres")
+	require.NoError(t, err)
+	require.NoError(t, psql.Close(ctx))
+}
+
+// TestProxyProtocolMySQL ensures that clients can successfully connect to a
+// MySQL database when Teleport is running behind a proxy that sends a proxy
+// line.
+func TestProxyProtocolMySQL(t *testing.T) {
+	ctx := context.Background()
+	testCtx := setupTestContext(ctx, t, withSelfHostedMySQL("mysql"))
+	t.Cleanup(func() { testCtx.Close() })
+	go testCtx.startHandlingConnections()
+
+	testCtx.createUserAndRole(ctx, t, "alice", "admin", []string{"root"}, []string{types.Wildcard})
+
+	// Point our proxy to the Teleport's MySQL listener.
+	proxy, err := newTestProxy(testCtx.mysqlListener.Addr().String())
+	require.NoError(t, err)
+	t.Cleanup(func() { proxy.Close() })
+	go proxy.Serve()
+
+	// Connect to the proxy instead of directly to MySQL listener and make
+	// sure the connection succeeds.
+	mysql, err := testCtx.mysqlClientAddr(proxy.Address(), "alice", "mysql", "root")
+	require.NoError(t, err)
+	require.NoError(t, mysql.Close())
+}
+
+// testProxy is tcp passthrough proxy that sends a proxy-line when connecting
+// to the target server.
+type testProxy struct {
```

Comment:
optional: move this to `lib/multiplexer`, in case other tests need a mock proxy-line implementation